### ISSUE TITLE
Get new page instance on each page visit

### ISSUE THREAD

**Issue by Austin-Castillo** [JByfordRew/BlazorTransitionableRoute#2]
Hello, 

Thank you for providing such a useful library. I had a question about the ability to transition/animate between pages while still getting a new instance of each page. 

As it is now, when you navigate from one page to another, it appears the 'secondary' page gets kept in memory. If you navigate back to it (so it is now the 'primary' page again), no lifecycle methods will fire for that page, and any private members will retain their old value.

You can see this in either of the demo apps. Navigate to the Counter page, increment the count, navigate to Home page, then back to Counter page. The count on the Counter page will be left with the old value. This is different than default Blazor navigation behavior, which will scrap the Counter page when you navigate away from it, and then initialize a new instance of it when you return to it.

In my case, I am relying on the page hitting those lifecycle methods on each visit to the page. Is there a way I can use this library to transition/animate between pages, while still keeping the default Blazor behavior of getting a new page instance on each visit?

Fyi, my app is a server application using Blazor coded transition behavior, if that matters at all.

Thanks for any help you are able to provide.

**Comment by JByfordRew** [comment]
Hi, many thanks for getting in touch and raising this. I've had a quick look but need to look into this some more. When I developed this it was for my use case which did require remembering state when navigating. I appreciate the default behaviour of Blazor and I am keen to implement an option to enable this. I hope to resolve this shortly. It won't be a breaking change but you will have to use the new option to choose the Blazor default behaviour.

**Comment by JByfordRew** [comment]
One thing to consider is that when transitioning back and depending on your style of transition and where your scroll position is may affect the integrity of the transition of the page gets recreated. This is an issue sometimes for navigating backwards which was why it was implemented this way for my use case.

**Comment by JByfordRew** [comment]
To help understand why it is like this. This is a snippet from the 'how it works' section of the read me file. I will provide an option to remedy though as soon as I can.

> Having two view layouts means we can remember the previous route but we cannot simply overwrite them, they must be preserved so we do not lose browser state like scroll position etc.

**Comment by JByfordRew** [comment]
The fundamental problem is to be able to transition you need to keep two RouteData alive to transition to.  As long as you move in one direction then the behaviour is as the Blazor default, forwards or backwards, but when you switch direction you hit your issue.  This is unfortunately core as to how this library functions.

So we need to trick the transition to believe it is always moving in one direction but recognises the browser back and forward events also to apply the correct transitional direction animation.

You may be able to use the value of the Transition.RouteData property to observe through OnParametersSet event and then fire a StateHasChanged call.

You may be able to modify you components to work slightly differently to move the control more to your code. It may be possible.

Failing this you may be able to make your URLs unique on each call perhaps.

Sorry I cannot be of more help until I have a more permanent solution

**Comment by Austin-Castillo** [comment]
Hi, thanks for relying so fast. I think in my particular case, losing things like the scroll position on the last page would be worth the tradeoff, thought I appreciate your point. 

For now I will try to use the value of Transistion.RouteData and see if I can trigger a StateHasChanged at the right time. Hopefully that will let me execute the code I need in the OnAfterRender for the re-visited page.

Please let me know if you think there's something else I should try or if you figure a way to solve it. Thanks!

**Comment by drew-fierst** [comment]
Hi, I too am  interested if you can find a solution to this issue. I chose to use your fantastic library because it is the only way I have found to be able to animate router transitions in Blazor, and the persistence of the state is actually an unwanted feature in my case.

For example, one route displays a component with a list of entities. Clicking on one of them transitions to another route where the entity is edited. When saved, it transitions back to the original route with the list of entities. However, since the list component's state was maintained, the user's edits are not reflected in the display. And, unfortunately, there's not even a lifecycle event that I can use to requery for the updated data.

Thank you for creating this library and I'd love to hear any ideas you have to resolve my dilemma.

**Comment by JByfordRew** [comment]
Hi, I have read your message and will look again. There is no immediate solution coming to mind. I do understand the importance for you both and others. Hopefully I can come up with something.

**Comment by JByfordRew** [comment]
Hi, I have tried to integrate this into the workings of the library but this is not easy to do.  The only solution I can find is to do the following.

For each of your page components that need this behaviour to reset per transition add `@inherits TransitionableLayoutComponent` to the top of the file and then implement a `protected override async Task OnParametersSetAsync()` and asynchronously delay a short while, about the length of the transition, before you reset the desired state of your component.  Here is the simple `Counter` page with this implementation.

```
@inherits TransitionableLayoutComponent

@page "/counter"

<h1>Counter</h1>

<p>Current count: @currentCount</p>

<button class="btn btn-primary" @onclick="IncrementCount">Click me</button>

@code {
    private int currentCount = 0;

    private void IncrementCount()
    {
        currentCount++;
    }

    protected override async Task OnParametersSetAsync()
    {
        await base.OnParametersSetAsync();

        if (!Transition.IntoView)
        {
            await Task.Delay(1000);
            ResetPage();
        }
    }

    private void ResetPage()
    {
        currentCount = 0;
    }
}
```

So the behaviour now is for routes that are transitioning out of view, after the transition completes or longer, you can safely reset the state without observing it during transition or thereafter.  

Taking this further to be more reusable you could create your own base component to implement the bulk of this behaviour and have an internal property that is an Action or Task that you could set with your reset logic.  This would mean you would only need to inherit from it and set the reset method on component initialisation.

I have also added this as a roadmap item to do but will close this issue now.

You may see a new inheritable class that does this as a optional fix in the library in the future but for now this is simple to do and should solve the issue.  

Let me know what you think and if it is appropriate for you?

**Comment by JByfordRew** [comment]
Hello all, I've implemented a very simple solution to the library that will solve this issue. Reopening issue until release.

**Comment by JByfordRew** [comment]
Fixed in V2.1.0

**Comment by drew-fierst** [comment]
Fantastic! Your modifications in version 2.1 are EXACTLY what I needed. Thank you!!
